File: src/main/java/com/timepath/steam/io/DiffGen.java
package com.timepath.steam.io;

import com.timepath.steam.io.storage.ACF;
import com.timepath.vfs.SimpleVFile;
import com.timepath.vfs.VFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.util.Collection;
import java.util.logging.Logger;

/**
 * @author TimePath
 */
class DiffGen {

    private static final Logger LOG = Logger.getLogger(DiffGen.class.getName());
    private static final String[] blacklist = {
            ".*/bin",
            ".*/cache",
            ".*tf/custom",
            ".*tf/download",
            ".*tf/replay",
            ".*tf/screenshots",
            ".*sounds?\\.cache",
            ".*cfg/config\\.cfg",
            ".*media/viewed\\.res",
            ".*tf/console\\.log",
            ".*tf/condump.*\\.txt",
            ".*tf/demoheader\\.tmp",
            ".*tf/voice_ban\\.dt",
            ".*tf/trainingprogress\\.txt"
    };
    private static final int K = 1024, M = K * K, G = M * K;

    private DiffGen() {
    }

    private static boolean check(@NotNull File f) {
        @NotNull String path = f.getPath();
        for (@NotNull String r : blacklist) {
            if (path.matches(r)) {
                LOG.info(path);
                return false;
            }
        }
        return true;
    }

    private static void extract(@NotNull VFile<? extends VFile> v, File dir) throws IOException {
        @NotNull File out = new File(dir, v.getName());
        if (!check(out)) {
            return;
        }
        if (v.isDirectory()) {
            out.mkdir();
            for (@NotNull VFile e : v.list()) {
                extract(e, out);
            }
        } else {
            out.createNewFile();
            @Nullable InputStream is = v.openStream();
            @NotNull FileOutputStream fos = new FileOutputStream(out);
            @NotNull BufferedOutputStream os = new BufferedOutputStream(fos);
            @NotNull byte[] buf = new byte[8 * M]; // r/w buffer
            int read;
            while ((read = is.read(buf)) >= 0) {
                os.write(buf, 0, read);
            }
            os.flush();
            os.close();
        }
    }

    public static void main(String... args) throws IOException {
        @NotNull int[] apps = {440};
        @NotNull File container = new File(System.getProperty("user.home"), "steamtracker");
        for (int i : apps) {
            @NotNull File repo = new File(container, String.valueOf(i));
            repo.mkdirs();
            @Nullable ACF acf = ACF.fromManifest(i);
            @Nullable Collection<? extends SimpleVFile> files = acf.list();
            for (@NotNull SimpleVFile v : files) {
                extract(v, repo);
            }
        }
        LOG.info("EXTRACTED");
    }
}

File: src/main/kotlin/com/timepath/steam/io/DiffGen.kt
package com.timepath.steam.io

import com.timepath.steam.io.storage.ACF
import com.timepath.vfs.VFile
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.logging.Logger
import kotlin.platform.platformStatic

public class DiffGen private constructor() {
    companion object {

        private val LOG = Logger.getLogger(javaClass<DiffGen>().getName())
        private val blacklist = arrayOf(".*/bin", ".*/cache", ".*tf/custom", ".*tf/download", ".*tf/replay", ".*tf/screenshots", ".*sounds?\\.cache", ".*cfg/config\\.cfg", ".*media/viewed\\.res", ".*tf/console\\.log", ".*tf/condump.*\\.txt", ".*tf/demoheader\\.tmp", ".*tf/voice_ban\\.dt", ".*tf/trainingprogress\\.txt")
        private val K = 1024
        private val M = K * K
        private val G = M * K

        private fun check(f: File): Boolean {
            val path = f.getPath()
            for (r in blacklist) {
                if (path.matches(r.toRegex())) {
                    LOG.info(path)
                    return false
                }
            }
            return true
        }

        throws(IOException::class)
        private fun <V : VFile<V>> extract(v: V, dir: File) {
            val out = File(dir, v.name)
            if (!check(out)) {
                return
            }
            if (v.isDirectory) {
                out.mkdir()
                for (e in v.list()) {
                    extract(e, out)
                }
            } else {
                out.createNewFile()
                v.openStream()!!.copyTo(FileOutputStream(out).buffered())
            }
        }

        throws(IOException::class)
        public platformStatic fun main(args: Array<String>) {
            val apps = intArrayOf(440)
            val container = File(System.getProperty("user.home"), "steamtracker")
            for (i in apps) {
                val repo = File(container, i.toString())
                repo.mkdirs()
                val acf = ACF.fromManifest(i)
                val files = acf.list()
                for (v in files) {
                    extract(v, repo)
                }
            }
            LOG.info("EXTRACTED")
        }
    }
}
